tests/components/http/test_forwarded.py

```diff
@@ -0,0 +1,487 @@
+        "/",
+        headers=[
+            (X_FORWARDED_FOR, "222.222.222.222"),
+            (X_FORWARDED_PROTO, "https"),
+            (X_FORWARDED_PROTO, "http"),
+        ],
+    )
+
+    assert resp.status == 400
+    assert "Too many headers for X-Forward-Proto" in caplog.text
+
+
+@pytest.mark.parametrize(
+    "x_forwarded_proto", ["", ",", "https, , https", "https, https, "],
+)
+async def test_x_forwarded_proto_empty_element(
+    x_forwarded_proto, aiohttp_client, caplog
+):
+    """Test that we get a HTTP 400 bad request with empty proto."""
+    app = web.Application()
+    app.router.add_get("/", mock_handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/", headers={X_FORWARDED_FOR: "1.1.1.1", X_FORWARDED_PROTO: x_forwarded_proto},
+    )
+
+    assert resp.status == 400
+    assert "Empty item received in X-Forward-Proto header" in caplog.text
+
+
+@pytest.mark.parametrize(
+    "x_forwarded_for,x_forwarded_proto,expected,got",
+    [
+        ("1.1.1.1, 2.2.2.2", "https, https, https", 2, 3),
+        ("1.1.1.1, 2.2.2.2, 3.3.3.3, 4.4.4.4", "https, https, https", 4, 3),
+    ],
+)
+async def test_x_forwarded_proto_incorrect_number_of_elements(
+    x_forwarded_for, x_forwarded_proto, expected, got, aiohttp_client, caplog
+):
+    """Test that we get a HTTP 400 bad request with incorrect number of elements."""
+    app = web.Application()
+    app.router.add_get("/", mock_handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/",
+        headers={
+            X_FORWARDED_FOR: x_forwarded_for,
+            X_FORWARDED_PROTO: x_forwarded_proto,
+        },
+    )
+
+    assert resp.status == 400
+    assert (
+        f"Incorrect number of elements in X-Forward-Proto. Expected 1 or {expected}, got {got}"
+        in caplog.text
+    )
+
+
+async def test_x_forwarded_host_without_trusted_proxy(aiohttp_client):
+    """Test that host header is ignored when untrusted."""
+
+    async def handler(request):
+        url = mock_api_client.make_url("/")
+        assert request.host == f"{url.host}:{url.port}"
+        assert request.scheme == "http"
+        assert not request.secure
+        assert request.remote == "127.0.0.1"
+
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+
+    async_setup_forwarded(app, [])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/",
+        headers={X_FORWARDED_FOR: "255.255.255.255", X_FORWARDED_HOST: "example.com"},
+    )
+
+    assert resp.status == 200
+
+
+async def test_x_forwarded_host_with_trusted_proxy(aiohttp_client):
+    """Test that we get the host header if proxy is trusted."""
+
+    async def handler(request):
+        assert request.host == "example.com"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)